pkg/jobworker/worker.go

```diff
@@ -0,0 +1,208 @@
+type ResourceController interface {
+	AddProcess(string, *exec.Cmd) error
+	CreateGroup(string) error
+	DeleteGroup(string) error
+	AddResourceControl(string, JobOpts) error
+}
+
+// NewOpts returns a configured JobOpts based on arguments
+func NewOpts(weight, ioLatency int32, memLimit CgroupByte) JobOpts {
+	return JobOpts{
+		CpuWeight: weight,
+		IOWeight:  ioLatency,
+		MemLimit:  memLimit,
+	}
+}
+
+// NewJob initialises a Job
+func NewJob(id string, cmd *exec.Cmd, con ResourceController) *Job {
+	return &Job{ID: id, running: true, cmd: cmd, con: con}
+}
+
+// Start calls start using the default ResourceController Cgroup
+func Start(opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	return StartWithController(&Cgroup{"/sys/fs/cgroup"}, opts, cmd, args...)
+}
+
+// StartWithController creates a job's cgroup, adds the resource controls from opts, creates a log file for the cgroup and
+// set's it to the exec.Cmd STDOUT and STDERR. Finally we Start the exec.Cmd and start a go routine to handle the blocking
+// call to Wait so that we can update the job's running flag.
+func StartWithController(con ResourceController, opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	// Create the job
+	j = NewJob(uuid.New().String(), exec.Command(cmd, args...), con)
+	// Create the cgroup and configure the controllers
+	if err = j.con.CreateGroup(j.ID); err != nil {
+		return nil, fmt.Errorf("failed to create cgroup: %w", err)
+	}
+	// Update cgroup controllers to add resource control to process
+	if err = j.con.AddResourceControl(j.ID, opts); err != nil {
+		return nil, fmt.Errorf("failed to add resource control: %w", err)
+	}
+	// Add job's process to cgroup
+	if err = j.con.AddProcess(j.ID, j.cmd); err != nil {
+		return nil, fmt.Errorf("failed to add PID to cgroup: %w", err)
+	}
+	defer syscall.Close(j.cmd.SysProcAttr.CgroupFD)
+	// Don't inherit environment from parent
+	j.cmd.Env = []string{}
+	// Pipe STDOUT and STDERR to a log file
+	f, err := os.OpenFile(logPath(j.ID), os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open job's log file: %w", err)
+	}
+	j.cmd.Stdout = f
+	j.cmd.Stderr = f
+	// Run the command as a given user as not to escalate privilege, since the executing user must also manage cgroups
+	if WORKER_UID != -1 && WORKER_GUID != -1 {
+		j.cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(WORKER_UID), Gid: uint32(WORKER_GUID)}
+	}
+	// Start the job
+	err = j.cmd.Start()
+	if err != nil {
+		return nil, fmt.Errorf("failed to start job's exec.Cmd: %w", err)
+	}
+	// Run go routine to handle the blocking call exec.Cmd.Wait() and update the running flag to indicate the job has complete
+	go func(runningJob *Job, logFile *os.File) {
+		runningJob.cmd.Wait()
+		runningJob.setRunning(false)
+		logFile.Close()
+	}(j, f)
+	return j, nil
+}
+
+// Stop request a job's termination using SIGTERM and deletes it's cgroup. If the SIGTERM is ignored, we send
```

Comment:
How are child processes handled? What if the command spawns a long-running child process and quits?

Reply:
The linux command get's executed as a process within a process group. Any child processes will be in the same group. This can be validated by running `sudo ./example bash -c "stress --cpu 2"` then checking the PGID with `ps -o pgid=`. It should be the PID of the job's command.

I just tested using `sudo ./example bash -c "stress --cpu 2 &"` though and realised my Stop function did not work as expected, since the PGID could not be obtained from the job in Stop since the command already quit. This left child processes running.

In my latest commit I fix this by getting a reference to the PGID when I start the job and use this later in Stop to signal the process group. You can test this by performing  `sudo ./example bash -c "stress --cpu 2 &"` then Ctrl+C and ensure no child processes are left over using `ps aux | grep stress`

Reply:
Does that child process stop gracefully?

Reply:
Yes, since we create a process group for the job all child processes are a part of this group, verified by the commands mentioned above. When we go to `Stop` the job, the SIGTERM sent will be sent to the process group and therefore sent to each process in that group.